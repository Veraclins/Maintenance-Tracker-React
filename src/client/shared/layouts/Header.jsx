import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';

import logo from '../../img/veratech-logo_burned.png';
import { logoutUser } from '../../features/authentication/authAction';
import NavItem from './NavItem';

export class Header extends Component {
  constructor() {
    super();
    this.state = {
      navbarActive: false,
      activeSubMenu: '',
    };
  }

  toggleNavbar = () => {
    const { navbarActive } = this.state;
    this.setState({ navbarActive: !navbarActive });
  }

  openSubMenu = (e) => {
    this.setState({ activeSubMenu: e.target.id });
  }

  closeNavbar = () => this.setState({ activeSubMenu: '', navbarActive: false });

  guestNav = () => (
    <React.Fragment>
      <NavItem closeNavbar={this.closeNavbar} link="/login">
        Login
      </NavItem>
      <NavItem closeNavbar={this.closeNavbar} link="/signup">
        Register
      </NavItem>
    </React.Fragment>
  )

  loggedInNav = () => {
    const { activeSubMenu } = this.state;
    const subMenuActiveClass = 'has-submenu active_submenu';
    const subMenuClass = 'has-submenu';
    return (
      <React.Fragment>
        <NavItem closeNavbar={this.closeNavbar} link="/dashboard">
          Dashboard
        </NavItem>
        <li
          className={activeSubMenu === 'request-menu-toggle'
            ? subMenuActiveClass
            : subMenuClass
          }
        >
          <a
            id="request-menu-toggle"
            onClick={this.openSubMenu}
          >
            Requests
          </a>
          <span className="arrow" />
          <div className="clear" />
          <ul className="sub-menu">
            <NavItem closeNavbar={this.closeNavbar} link="/requests/create">
            Create a Request
            </NavItem>
            <NavItem closeNavbar={this.closeNavbar} link="/requests">
            View Requests
            </NavItem>
          </ul>
        </li>
      </React.Fragment>
    );
  }

  navigation = () => {
    const { navbarActive } = this.state;
    const { logout, isLoggedIn, isAdmin } = this.props;
    return (
      <div className="navigation">
        <a
          className={navbarActive
            ? 'nav_toggle active'
            : 'nav_toggle'}
          onClick={this.toggleNavbar}
        >
          <span />
          <span />
          <span />
          <span />
        </a>
        <div className={navbarActive ? 'menu-top-menu-container show' : 'menu-top-menu-container'}>
          <ul id="top-menu" className="menu">
            {isLoggedIn ? this.loggedInNav() : false}
            {isAdmin ? (
              <NavItem closeNavbar={this.closeNavbar} link="/admin">
                Admin Page
              </NavItem>
            ) : false}
            {!isLoggedIn ? this.guestNav() : (
              <li className="menu-item">
                <a
                  className="buttons button_default"
                  onClick={() => { this.closeNavbar(); logout(); }}
                >
                  Logout
                </a>
                <span className="arrow" />
                <div className="clear" />
              </li>
            )}
          </ul>
        </div>
      </div>
    );
  };

  render() {
    return (
      <header>
        <div className="header_layout">
          <div className="container">
            <div className="logo">
              <NavLink to="/">
                <img
                  src={logo}
                  alt="VeraTech"
                  width="144"
                  height="42"
                />
              </NavLink>
            </div>
            {this.navigation()}
          </div>
        </div>
      </header>
    );
  }
}

Header.propTypes = {
  logout: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  isAdmin: PropTypes.bool.isRequired,
};

export const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logoutUser()),
});

export const mapStateToProps = state => ({
  isLoggedIn: state.auth.isAuthenticated,
  isAdmin: state.auth.isAdmin,
});
export default connect(mapStateToProps, mapDispatchToProps)(Header);
