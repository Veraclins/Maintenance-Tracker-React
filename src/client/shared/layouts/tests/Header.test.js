import React from 'react';
import { shallow } from 'enzyme';
import { NavLink } from 'react-router-dom';

import { Header, mapDispatchToProps } from '../Header';

const mockFunction = jest.fn();

describe('Test the Header component', () => {
  it('renders without crashing', () => {
    const isAdmin = false;
    const isLoggedIn = false;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    expect(wrapper.find('header')).toHaveLength(1);
    expect(wrapper.find('img')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the right nav elements for a guest', () => {
    const isAdmin = false;
    const isLoggedIn = false;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    expect(wrapper.find(NavLink)).toHaveLength(3);
    expect(wrapper.find(NavLink).get(1).props.children).toEqual('Login');
    expect(wrapper.find(NavLink).get(2).props.children).toEqual('Register');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the right nav elements for a logged in user', () => {
    const isAdmin = false;
    const isLoggedIn = true;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    expect(wrapper.find(NavLink)).toHaveLength(4);
    expect(wrapper.find(NavLink).get(1).props.children).toEqual('Dashboard');
    expect(wrapper.find(NavLink).get(2).props.children).toEqual('Create a Request');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the right nav elements for an admin', () => {
    const isAdmin = true;
    const isLoggedIn = true;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    expect(wrapper.find(NavLink)).toHaveLength(5);
    expect(wrapper.find(NavLink).get(1).props.children).toEqual('Dashboard');
    expect(wrapper.find(NavLink).get(4).props.children).toEqual('Admin Page');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the right nav elements for an admin', () => {
    const isAdmin = true;
    const isLoggedIn = true;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    expect(wrapper.find('.button_default')).toHaveLength(1);
    wrapper.find('.button_default').simulate('click');
    expect(mockFunction).toHaveBeenCalled();
  });
});

describe('directly invoking the toggleNavbar method from component instance', () => {
  it('should update the navbarActive when invoked', () => {
    const isAdmin = false;
    const isLoggedIn = true;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    const instance = wrapper.instance();
    expect(wrapper.state('navbarActive')).toEqual(false);
    instance.toggleNavbar();
    expect(wrapper.state('navbarActive')).toEqual(true);
  });
});

describe('directly invoking the openSubMenu method from component instance', () => {
  it('should update the activeSubMenu when invoked', () => {
    const isAdmin = false;
    const isLoggedIn = true;
    const wrapper = shallow(<Header
      isAdmin={isAdmin}
      isLoggedIn={isLoggedIn}
      logout={mockFunction}
    />);
    const e = {
      target: {
        id: 'nav-1',
      },
    };
    const instance = wrapper.instance();
    expect(wrapper.state('activeSubMenu')).toEqual('');
    instance.openSubMenu(e);
    expect(wrapper.state('activeSubMenu')).toEqual('nav-1');
  });
});

describe('Testing mapDispatchToProps', () => {
  it('should correctly map dispatches to props  ', () => {
    const mockDispatch = jest.fn();
    const newProps = mapDispatchToProps(mockDispatch);
    newProps.logout();
    expect(mockDispatch)
      .toHaveBeenCalledTimes(1);
  });
});
