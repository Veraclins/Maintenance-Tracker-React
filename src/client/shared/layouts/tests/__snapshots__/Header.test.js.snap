// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Header component renders the right nav elements for a guest 1`] = `
<header>
  <div
    className="header_layout"
  >
    <div
      className="container"
    >
      <div
        className="logo"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/"
        >
          <img
            alt="VeraTech"
            height="42"
            src="test-file-stub"
            width="144"
          />
        </NavLink>
      </div>
      <div
        className="navigation"
      >
        <a
          className="nav_toggle"
          onClick={[Function]}
        >
          <span />
          <span />
          <span />
          <span />
        </a>
        <div
          className="menu-top-menu-container"
        >
          <ul
            className="menu"
            id="top-menu"
          >
            <NavItem
              closeNavbar={[Function]}
              link="/login"
            >
              Login
            </NavItem>
            <NavItem
              closeNavbar={[Function]}
              link="/signup"
            >
              Register
            </NavItem>
          </ul>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`Test the Header component renders the right nav elements for a logged in user 1`] = `
<header>
  <div
    className="header_layout"
  >
    <div
      className="container"
    >
      <div
        className="logo"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/"
        >
          <img
            alt="VeraTech"
            height="42"
            src="test-file-stub"
            width="144"
          />
        </NavLink>
      </div>
      <div
        className="navigation"
      >
        <a
          className="nav_toggle"
          onClick={[Function]}
        >
          <span />
          <span />
          <span />
          <span />
        </a>
        <div
          className="menu-top-menu-container"
        >
          <ul
            className="menu"
            id="top-menu"
          >
            <NavItem
              closeNavbar={[Function]}
              link="/dashboard"
            >
              Dashboard
            </NavItem>
            <li
              className="has-submenu"
            >
              <a
                id="request-menu-toggle"
                onClick={[Function]}
              >
                Requests
              </a>
              <span
                className="arrow"
              />
              <div
                className="clear"
              />
              <ul
                className="sub-menu"
              >
                <NavItem
                  closeNavbar={[Function]}
                  link="/requests/create"
                >
                  Create a Request
                </NavItem>
                <NavItem
                  closeNavbar={[Function]}
                  link="/requests"
                >
                  View Requests
                </NavItem>
              </ul>
            </li>
            <li
              className="menu-item"
            >
              <a
                className="buttons button_default"
                onClick={[Function]}
              >
                Logout
              </a>
              <span
                className="arrow"
              />
              <div
                className="clear"
              />
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`Test the Header component renders the right nav elements for an admin 1`] = `
<header>
  <div
    className="header_layout"
  >
    <div
      className="container"
    >
      <div
        className="logo"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/"
        >
          <img
            alt="VeraTech"
            height="42"
            src="test-file-stub"
            width="144"
          />
        </NavLink>
      </div>
      <div
        className="navigation"
      >
        <a
          className="nav_toggle"
          onClick={[Function]}
        >
          <span />
          <span />
          <span />
          <span />
        </a>
        <div
          className="menu-top-menu-container"
        >
          <ul
            className="menu"
            id="top-menu"
          >
            <NavItem
              closeNavbar={[Function]}
              link="/dashboard"
            >
              Dashboard
            </NavItem>
            <li
              className="has-submenu"
            >
              <a
                id="request-menu-toggle"
                onClick={[Function]}
              >
                Requests
              </a>
              <span
                className="arrow"
              />
              <div
                className="clear"
              />
              <ul
                className="sub-menu"
              >
                <NavItem
                  closeNavbar={[Function]}
                  link="/requests/create"
                >
                  Create a Request
                </NavItem>
                <NavItem
                  closeNavbar={[Function]}
                  link="/requests"
                >
                  View Requests
                </NavItem>
              </ul>
            </li>
            <NavItem
              closeNavbar={[Function]}
              link="/admin"
            >
              Admin Page
            </NavItem>
            <li
              className="menu-item"
            >
              <a
                className="buttons button_default"
                onClick={[Function]}
              >
                Logout
              </a>
              <span
                className="arrow"
              />
              <div
                className="clear"
              />
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`Test the Header component renders without crashing 1`] = `
<header>
  <div
    className="header_layout"
  >
    <div
      className="container"
    >
      <div
        className="logo"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/"
        >
          <img
            alt="VeraTech"
            height="42"
            src="test-file-stub"
            width="144"
          />
        </NavLink>
      </div>
      <div
        className="navigation"
      >
        <a
          className="nav_toggle"
          onClick={[Function]}
        >
          <span />
          <span />
          <span />
          <span />
        </a>
        <div
          className="menu-top-menu-container"
        >
          <ul
            className="menu"
            id="top-menu"
          >
            <NavItem
              closeNavbar={[Function]}
              link="/login"
            >
              Login
            </NavItem>
            <NavItem
              closeNavbar={[Function]}
              link="/signup"
            >
              Register
            </NavItem>
          </ul>
        </div>
      </div>
    </div>
  </div>
</header>
`;
